redis/src/main/scala/zio/redis/TestExecutor.scala

```diff
@@ -96,6 +96,281 @@ private[redis] final class TestExecutor private (
       case api.Connection.Select.name =>
         onConnection(name, input)(RespValue.bulkString("OK"))
 
+      case api.Geo.GeoAdd =>
+        val key = input.head.asString
+
+        val values =
+          Chunk.fromIterator(
+            input.tail
+              .map(_.asString)
+              .grouped(3)
+              .map(g => (LongLat(g(0).toDouble, g(1).toDouble), g(2)))
+          )
+
+        orInvalidParameter(STM.succeed(values.map(_._1).forall(Hash.isValidLongLat)))(orWrongType(isSortedSet(key)) {
+          val members = values.map { case (longLat, member) =>
+            MemberScore(Hash.encodeAsHash(longLat.longitude, longLat.latitude).toDouble, member)
+          }
+
+          for {
+            scoreMap    <- sortedSets.getOrElse(key, Map.empty)
+            membersAdded = members.count(ms => scoreMap.contains(ms.member))
+            newScoreMap  = scoreMap ++ members.map(ms => (ms.member, ms.score)).toMap
+            _           <- sortedSets.put(key, newScoreMap)
+          } yield RespValue.Integer(membersAdded.toLong)
+        })
+
+      case api.Geo.GeoDist =>
+        val key = input.head.asString
+        val unit = input.last.asString match {
+          case "m"  => Some(RadiusUnit.Meters)
+          case "km" => Some(RadiusUnit.Kilometers)
+          case "ft" => Some(RadiusUnit.Feet)
+          case "mi" => Some(RadiusUnit.Miles)
+          case _    => None
+        }
+        val members = (if (unit.isEmpty) input.tail else input.dropRight(1)).map(_.asString)
+
+        orWrongType(isSortedSet(key))(
+          for {
+            scoreMap <- sortedSets.getOrElse(key, Map.empty)
+            hashes    = members collect scoreMap
+          } yield
+            if (hashes.size == 2)
+              RespValue.bulkString(
+                Hash
+                  .distance(
+                    Hash.decodeHash(hashes(0).toLong),
+                    Hash.decodeHash(hashes(1).toLong),
+                    unit.getOrElse(RadiusUnit.Meters)
+                  )
+                  .toString
+              )
+            else RespValue.NullBulkString
+        )
+
+      case api.Geo.GeoHash =>
+        val key     = input.head.asString
+        val members = input.tail.map(_.asString)
+
+        orWrongType(isSortedSet(key))(
+          for {
+            scoreMap <- sortedSets.getOrElse(key, Map.empty)
+            hashes    = members.map(scoreMap.get)
+          } yield RespValue.Array(
+            hashes.map { hashOption =>
+              hashOption.map { hash =>
+                val geoHash = Hash.asGeoHash(hash.toLong)
+                RespValue.bulkString(geoHash)
+              }.getOrElse(RespValue.NullBulkString)
```

Comment:
Prefer `fold` to `map` + `getOrElse`. Less important, avoid long expressions in `yield`. You can calculate the array contents before.